python/lib/communication/dmod/communication/registered/registered_interface.py

```diff
@@ -0,0 +1,391 @@
+from .aliases import *
+
+
+class RegisteredWebSocketInterface(WebSocketInterface, abc.ABC):
+    """
+    A websocket interface implementation that routes logic through registered initializers, consumers, and producers
+
+    Initializers, consumers, producers, and handlers are usually defined via decorator
+    """
+    @classmethod
+    def get_parseable_request_types(cls) -> typing.List[typing.Type[AbstractInitRequest]]:
+        return list()
+
+    def __init__(
+        self,
+        listen_host: str = '127.0.0.1',
+        port: typing.Union[bytes, str, int] = 3012,
+        ssl_dir: str = None,
+        cert_pem: str = None,
+        priv_key_pem: str = None,
+        use_ssl: bool = True,
+        message_handlers: typing.Mapping[typing.Type[AbstractInitRequest], AbstractRequestHandler] = None,
+        *args,
+        **kwargs
+    ):
+        super().__init__(
+            listen_host=listen_host,
+            port=port,
+            ssl_dir=ssl_dir,
+            cert_pem=cert_pem,
+            priv_key_pem=priv_key_pem,
+            use_ssl=use_ssl,
+            *args,
+            **kwargs
+        )
+        print(f"Launching {self.__class__.__name__}, listening to {listen_host}:{str(port)}...")
+
+        self._event_handlers: typing.Dict[typing.Type[AbstractInitRequest], AbstractRequestHandler] = dict()
+        """A mapping between a type of message and the function that is supposed to consume it"""
+
+        if message_handlers:
+            for message_type, handler in message_handlers:
+                self.register_event_handler(message_type, handler)
+
+        # Run all recognized initialization functions
+        self._initialize(
+            listen_host=listen_host,
+            port=port,
+            ssl_dir=ssl_dir,
+            cert_pem=cert_pem,
+            priv_key_pem=priv_key_pem,
+            *args,
+            **kwargs
+        )
+        self._assign_handlers()
+
+    def _initialize(self, *args, **kwargs):
+        """
+        Calls all added initialization functions and attaches all consumers and producers
+        """
+        for initialization_function in self._get_initialization_functions():
+            initialization_function(*args, **kwargs)
+
+    def _get_initialization_functions(self) -> typing.Sequence[VARIABLE_CALLABLE]:
+        """
+        Generates a list of all member functions that need to be called at the end of the abstract class construction
+
+        Each function must handle `*args, **kwargs`, and the first arguments handled will be the arguments passed into
+        `RegistedWebSocketInterface` constructor.
+
+        Returns:
+            All member functions that need to be called at the end of the abstract class construction
+        """
+
+        initialization_functions = decorators.find_functions_by_decorator(
+            self,
+            decorators.initializer
+        )
+
+        initializers: typing.List[VARIABLE_CALLABLE] = list()
```

Comment:
Thanks for all the explicit type hints!

Reply:
That particular type hint will need to change in the future. I don't remember what version it is, but `typing` is going to support generic variable arguments (`*args` and `**kwargs`) at some point in the future. I think we have a ways to go, so this is just a hack.